### PR TITLE
ENH: add len to several places

### PR DESCRIPTION
**Pull request by adler-j**
See #424, but this adds it to several more places and follows numpy convention.

### CODE REVIEW

File: odl/discr/lp_discr.py [hunk 1]
```diff
@@ -188,6 +188,13 @@ def size(self):
         """Total number of underlying partition cells."""
         return self.partition.size
 
+    def __len__(self):
+        """Return ``len(self)``.
+
+        Total number of underlying partition cells.
+        """
+        return self.size
```
Comment (kohr-h) [inline odl/discr/lp_discr.py]: This should return `len(self.partition)`, no? `DiscreteLp` "looks like" the partition it's defined on.

File: odl/space/pspace.py [hunk 2]
```diff
@@ -305,6 +305,10 @@ def size(self):
         """The number of factors."""
         return self._size
 
+    def __len__(self):
+        """Return ``len(self)``."""
+        return self._size
```
Comment (kohr-h) [inline odl/space/pspace.py]: w/o underscore

**Comment by kohr-h** [comment]
Is this interpretation of the "length" of a discretization obvious? IMO, it is only "kind of" obvious. The alternative, which we should consider an option, is not to implement it at all for `Discretization` classes. For an element, on the other hand, the interpretation is very clear.

File: odl/discr/lp_discr.py (hunk 1 above)
Comment (adler-j) [inline odl/discr/lp_discr.py]: Actually, it does not. The `[ ]` operator of `DiscreteLp` is in fact linear.

**Comment by adler-j** [comment]
It is the only valid interpretation however, otherwise stuff like `list(vector)` will fail.

The only obvious fix (and something we need to do at some point I suppose) is to start supporting multi-indexes for `DiscreteLpVector`, then we could change that.

**Comment by adler-j** [comment]
Additional note:

I just did some checking and `__len__` already exists for `DiscreteLp`, (by inheritance from `BaseNtuple`)

**Comment by kohr-h** [comment]
> It is the only valid interpretation however, otherwise stuff like list(vector) will fail.

Vectors are perfectly fine, I was more concerned about the space itself.



> The only obvious fix (and something we need to do at some point I suppose) is to start supporting multi-indexes for DiscreteLpVector, then we could change that.

That would be the best outcome, but that depends on a supporting CUDA back-end I guess. But actually, you're right, `DiscreteLp` should be considered flat for now since multiple indices are not supported.

> I just did some checking and __len__ already exists for DiscreteLp, (by inheritance from BaseNtuple). It returns the same as size.

Then it's probably best to do nothing right now and inherit, and change behavior as soon as multi-indexing is available.

File: odl/discr/discretization.py [hunk 3]
```diff
@@ -242,16 +242,6 @@ def ntuple(self):
         """Structure for data storage."""
         return self._ntuple
 
-    @property
```
Comment (adler-j) [inline odl/discr/discretization.py]: These follow from inheritance

**Comment by adler-j** [comment]
Updated according to comments, merge?

**Comment by kohr-h** [comment]
LGTM

**Pull request merged**